### ISSUE TITLE
Instantiate client pointer

### ISSUE THREAD

**Issue by ptrsr** [eclipse/paho.mqtt.cpp#141]
Hello, I've been using the MQTT c++ library with success but I do not understand how the mqtt::client_ptr works.

I see that it's a shared pointer, but how do I instantiate it? the constructor is deleted, and std::make_shared<mqtt::client>() doesn't work as well.

**Comment by saumilagg** [comment]
hey @ptrsr ...did you find a solution? Its not working for me either. I am instantiating like this: 

mqtt::async_client_ptr clientPtr = std::make_shared<mqtt::async_client>(tempStr, sn);

tempStr and sn are std::string.

seg fault from remove_token at async_Client.cpp. A callback pointer is causing this.

**Comment by fpagliughi** [comment]
Yes, sorry. For the sake of consistency I put in a `ptr_t` definition for most of the classes and then an external *type_ptr* definition for each. But then I only added create methods for the really dynamic types like messages and tokens. 

There's nothing magic about the create functions; they just call `make_shared<>` for the type. So you should be able to use it with one of the constructors. The empty constructor is deleted because you need to specify, at a minimum, the server URI and client ID when creating the object. Those constructors _should_ work with a shared pointer, no problem.

The problem with shared pointers for clients is that you **must** keep one reference to the client alive at all times. And therefore I found it easier to just declare one on the stack at the top of main, and thus keep it alive for the entire run of the application.

@saumilagg could your shared pointer be going out of scope at some point, allowing the client to be deleted before the callback occurs? That would certainly cause a segfault as the callback would be attempting to dereference a pointer to garbage at that point.

**Comment by fpagliughi** [comment]
This gets a little tricky, On the one hand, @ptrsr , you can easily create a shared pointer to the client, but you must do so manually, like:
```
mqtt::client_ptr client = std::make_shared<mqtt::client>(SERVER_ADDRESS, CLIENT_ID);
```
or
```
mqtt::async_client_ptr client = std::make_shared<mqtt::async_client>(SERVER_ADDRESS, CLIENT_ID);
```

But, as of this time, the client doesn't take ownership of most of the objects handed to it. An update to the Paho C client will have it make a copy of the client options (#145), But it is not taking control of the callback or persistence objects, if they're used.

In the new Paho Rust library, I solved this issue by transferring ownership of these objects into the client. (Rust is much more explicit about ownership issues). In C++ we could do this by moving the objects into the client. 

@saumilagg When I tried to recreate this, I kept the client smart pointer alive, but crashed on the callback anyway. I realized that I had put the callback object on the stack, and it went out of scope. So when a callback arrived... segfault!

**Comment by saumilagg** [comment]
ya I was trying to debug it and check if the pointer is going out of scope. But as you said when a callback...segfault!! What do you think could we do for a solution? @fpagliughi 

I cannot have the client object in my class constructor as the arguments (SERVER_ADDRESS, CLIENT_ID) are populated after the widget/app loads and user inputs!

**Comment by fpagliughi** [comment]
You can use a shared pointer. You just need to keep the related objects alive for the life of the client. Perhaps you can create a struct or tuple that keeps a shared pointer to each of the things you need. Here's a hacked version of part of the `async_client` sample that uses a tuple:
```
using my_client = std::tuple<mqtt::async_client_ptr, 
                mqtt::connect_options_ptr, 
                mqtt::callback_ptr>;

my_client connect()
{
    auto connOpts = std::make_shared<mqtt::connect_options>();
    connOpts->set_keep_alive_interval(20);
    connOpts->set_clean_session(true);

    auto client = std::make_shared<mqtt::async_client>(SERVER_ADDRESS, CLIENT_ID);

    auto cb = std::make_shared<callback>(*client, *connOpts);
    client->set_callback(*cb);

    // Start the connection.
    try {
        std::cout << "Connecting to the MQTT server..." << std::flush;
        client->connect(*connOpts, nullptr, *cb);
    }
    catch (const mqtt::exception&) {
        std::cerr << "\nERROR: Unable to connect to MQTT server: '"
            << SERVER_ADDRESS << "'" << std::endl;
        exit(1);
    }

    return std::make_tuple(client, connOpts, cb);
}

/////////////////////////////////////////////////////////////////////////////

int main(int argc, char* argv[])
{
    auto clientTpl = connect();

    // Just block till user tells us to quit.

    while (std::tolower(std::cin.get()) != 'q')
        ;

    ...
}
```
It feels a bit hacky, but it works.

**Comment by saumilagg** [comment]
That is a good hack, thanks! Did it compile for you? @fpagliughi  I was working on it and have the callback error ( no matching function call to 'callback::callback(mqtt::async_client &, mqtt::connect_options &)' on file new_allocator.h 

Seems like some function pointer error? Still not able to fix it. Did you get any such error? @fpagliughi

**Comment by fpagliughi** [comment]
Yes, it worked fine for me. I compiled it on Ubuntu 16.04 with GCC 5.2, using the development branch of the Paho C++ library. But I imagine it should be generally workable.

**Comment by saumilagg** [comment]
Ok, checking out the dev branch. I did on master with centos 7 and gcc4.8

**Comment by saumilagg** [comment]
Got it. Works very well. 

Had a constructor issue. Thanks a lot sir @fpagliughi :)

Do you suggest working off dev branch?

**Comment by fpagliughi** [comment]
I would think that the example _should_ work off the v1.0 release or master branch. I just mentioned that I happened to be working with the _develop_ branch, and inadvertently used it to test the sample I posted above.

The  _develop_ branch has the latest features and bug fixes as I move to the next release, but it could be unstable. It all depends what you are trying to accomplish with your work, or when you expect to release your own code.
